```{r, echo=FALSE}
knitr::opts_chunk$set(cache=TRUE, error=FALSE, warning=FALSE, message=FALSE)
```

```{r loading}
library(scRNAseq)
sce.seger <- SegerstolpePancreasData()
```

```{r gene-annotation}
library(org.Hs.eg.db)
symbols <- rowData(sce.seger)$symbol
ens.id <- mapIds(org.Hs.eg.db, keys=symbols, keytype="SYMBOL", column="ENSEMBL")
ens.id <- ifelse(is.na(ens.id), symbols, ens.id)

# Removing duplicated rows.
keep <- !duplicated(ens.id)
sce.seger <- sce.seger[keep,]
rownames(sce.seger) <- ens.id[keep]
```

```{r sample-annotation}
# We simplify the names of some of the relevant 
# column metadata fields for ease of access.
emtab.meta <- colData(sce.seger)[,c("cell type", 
    "individual", "single cell well quality")]
colnames(emtab.meta) <- c("CellType", "Donor", "Quality")
colData(sce.seger) <- emtab.meta

# Some editing of the cell type labels is necessary 
# for consistency with other data sets.
sce.seger$CellType <- gsub(" cell", "", sce.seger$CellType)
sce.seger$CellType <- paste0(
    toupper(substr(sce.seger$CellType, 1, 1)),
    substring(sce.seger$CellType, 2))
```

```{r quality-control}
# Removing low quality cells that were marked by the authors.
low.qual <- sce.seger$Quality == "low quality cell"

# Performing additional quality control, just in case.
library(scater)
sce.seger <- calculateQCMetrics(sce.seger, compact=TRUE)
qc.mat <- cbind(
    NFeatures=isOutlier(sce.seger$scater_qc$all$total_features_by_counts, 
        log=TRUE, type="lower", nmads=3),
    LibSize=isOutlier(sce.seger$scater_qc$all$total_counts, 
        log=TRUE, type="lower", nmads=3),
    SpikePct=isOutlier(sce.seger$scater_qc$feature_control_ERCC$pct_counts, 
        type="higher", nmads=3)
)

discard <- rowMeans(qc.mat) > 0 | low.qual
sce.seger <- sce.seger[,!discard]
```

```{r normalization}
library(scran)
clusters <- quickCluster(sce.seger, BSPARAM=BiocSingular::IrlbaParam())
sce.seger <- computeSumFactors(sce.seger, clusters=clusters)
sce.seger <- computeSpikeFactors(sce.seger, general.use=FALSE)
```

```{r variance-modelling}
# Not using cells with no spike-ins for variance modelling
# (Donor AZ has very spike-ins and is subsequently ignored).
for.hvg <- sce.seger[,sizeFactors(sce.seger, "ERCC") > 0
    & sce.seger$Donor!="AZ"]
for.hvg <- multiBlockNorm(for.hvg, for.hvg$Donor) 
dec.seger <- multiBlockVar(for.hvg, for.hvg$Donor)
```
