---
output:
  html_document
bibliography: ../ref.bib
---

# Sharing and Interactive Interfaces {#sharing-interactive-ui}

```{r setup, echo=FALSE, results="asis"}
source("workflows/knitr_options.R")
```

## Motivation

A major part of science is to share data and observations with collaborators and the wider scientific community.
However, single-cell analysis workflows generate such abundant and diverse data that they cannot be comprehensively captured and reported in static reports (e.g., scientific articles).
As such, interactive graphical user interfaces (GUI) for exploratory data analysis have gained popularity in recent years to enable other scientists to explore a dataset independently from the findings reported in the primary analysis.
Notably, an [awesome repository](https://github.com/federicomarini/awesome-expression-browser) has been set up to collaboratively track interactive expression browsers as they become available.
In this chapter, we describe solutions for sharing single-cell genomics data using interactive browsers directly integrating with Bioconductor workflows and data structures.

## Quick Start

In this section we use the 10X PBMC dataset for demonstration purposes.

```{r, results='asis', echo=FALSE}
extractCached("workflows/tenx-pbmc4k", "dimensionality-reduction", "sce.pbmc")
```

```{r}
sce.pbmc
```

To share data and enable interactive exploration we recommend loading an `sce` object in the iSEE app via:

```{r}
library(iSEE)
if (interactive()) {
    iSEE(sce.pbmc)
}
```

For examples of the iSEE interface applied to various types of genomics datasets, see https://github.com/csoneson/iSEE, https://github.com/LTLA/iSEE2018, https://github.com/federicomarini/iSEE_instances.

To get started, iSEE apps each provide a guided tour to walk the user through the functionality of the interface.

For analysts generating the `sce` object, we recommend saving the `sce` via one of the following methods :

```{r save_rds, message=FALSE}
# method 1
tempfile1 <- tempfile(fileext = ".rds")
saveRDS(sce.pbmc, tempfile1)
# method 2
library(HDF5Array)
tempdir1 <- tempfile("h5_sce_")
sce.pbmc <- saveHDF5SummarizedExperiment(sce.pbmc, tempdir1)
```

To read a saved `sce` object back in, you can then run:

```{r}
# method 1
sce.pbmc <- readRDS(tempfile1)
# method 2
sce.pbmc <- loadHDF5SummarizedExperiment(tempdir1)
```

## Exploring SummarizedExperiment objects

The `SummarizedExperiment` class - extended by the `SingleCellExperiment` class - provides a flexible and powerful container to store information generated by single-cell RNA-sequencing workflows (see the [Data Infrastructure](#data-infrastructure) chapter for more details).

The `iSEE` package leverages the stable yet versatile structure of the `SummarizedExperiment` class to deliver an interactive GUI capable of displaying any genomics data stored in the various components of a `SummarizedExperiment` or `SingleCellExperiment` object.

In particular, the `iSEE` interface provides 8 different panel types, each focusing on a distinct aspect of the data and metadata stored in the various components of an `sce` object.
The panels are described in greater details in the [package vignettes](http://bioconductor.org/packages/release/bioc/html/iSEE.html) and the [workshop materials](https://github.com/kevinrue/iSEEWorkshop2019) presented at the Bioconductor 2019 conference.

## Advanced Usage

Conveniently, individual `iSEE` apps can be preconfigured to launch the interface in any predefined state.
This is particularly useful to showcase certain aspects of a dataset, and save time once users have identified a configuration that fits their particular dataset and purpose.

Furthermore, the `iSEE` app interface can accomodate user-defined custom panels.
Custom panels can be plots and tables that are generated by user-defined functions applied to selections received from other panels in the interface.
For examples of custom panels, see https://github.com/kevinrue/iSEE_custom.

Instances of `iSEE` apps can be hosted publicly or privately using a [Shiny server](https://www.rstudio.com/products/shiny/shiny-server/).

Finally, the `iSEE` interface is capable of tracking and displaying at any time the code necessary to reproduce all the panels visible in the interface.
This feature is particularly useful for teaching purposes and reproducibity.

## Session Info

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
```
