# Sharing Data and Interactive Interfaces {#sharing-data-ui}

```{r setup, include=FALSE, results='asis'}
source("workflows/extractor.R")
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE)
library(BiocStyle)
setupHTML()
```

## Motivation

As demonstrated in other chapters of this book, single-cell analysis workflows rapidly generate a huge amount of diverse data---all the way from raw experimental data and metadata to the result of complex workflows---, far exceeding what can be interactively and thoroughly examined by any individual scientist.
Therefore, the capacity to share access to raw and processed data is key to making the most of any dataset.

While predefined static reports generated by automated workflows can be used to rapidly summarize and visualize common aspects of a datasets (e.g., quality control metrics), interactive data exploration is critical to the analysis and comprehension and interpretation of individual datasets.
In particular, exploration drives the formation of novel data-driven hypotheses prior to a more rigorous statistical analysis, and enables diagnosis of potential problems such as batch effects and low-quality samples.

Moreover, exposing a dataset to scrutiny by collaborators and the wider scientific community not only facilitates and accelerates the discovery process, but also promotes trust in transparent and reproducible research.

## iSEE: The interactive SummarizedExperiment Explorer

The `SummarizedExperiment` class---extended by the `SingleCellExperiment` class---provides a flexible and powerful container to store information generated by single-cell RNA-sequencing workflows (see the [Data Infrastructure](#data-infrastructure) chapter for more details).

![Overview of the *iSEE* interactive data visualization package's relationship to a *SingleCellExperiment* object](images/iSEE-dataflow.png)

For demonstration, we will use the 10X PBMC dataset:

```{r, echo=FALSE, results="asis"}
extractCached("workflows/tenx-pbmc4k", "normalization", "sce.pbmc")
```
