---
output:
  html_document
bibliography: ../ref.bib
---

# Integrating with protein abundance

```{r setup, echo=FALSE, results="asis"}
library(OSCAUtils)
chapterPreamble()
```

```{r, fig.cap="This page is under construction. Enjoy these platelets at work instead.", echo=FALSE}
knitr::include_graphics("https://honeysanime.com/wp-content/uploads/2018/10/Hataraku-Saibou-Cells-at-Work-Wallpaper-5.jpg")
```

## Motivation

Cellular indexing of transcriptomes and epitopes by sequencing (CITE-seq) is a technique that quantifies both gene expression and the abundance of selected surface proteins in each cell simultaneously [@stoeckius2017simultaneous].
In this approach, cells are first labelled with antibodies that have been conjugated with synthetic RNA tags.
A cell with a higher abundance of a target protein will be bound by more antibodies, resulting in more of the corresponding tags.
Both the tags and endogenous transcripts are reverse transcribed and captured in a cDNA library; the abundance of each protein or expression of each gene is subsequently quantified by sequencing of each set of features.
This provides a powerful tool for interrogating aspects of the proteome (such as post-translational modifications) and other cellular features that would normally be invisible to transcript-based studies.

How should the data for these antibody tags be incorporated into the analysis?
While we have counts for both tags and transcripts, there are fundamental differences in nature of the data that make it difficult to treat the former as additional features in the latter.
Most experiments involve only a small number of antibodies (<20) that are chosen by the researcher because they are of _a priori_ interest, in contrast to gene expression data that captures the entire transcriptome regardless of the study.
The coverage of the tags is also much deeper as they are sequenced separately from the transcripts, allowing the sequencing resources to be concentrated into a smaller number of features.
And, of course, the use of antibodies against protein targets involves consideration of separate biases compared to those observed for transcripts.

In this chapter, we will describe some strategies for integrated analysis of antibody tag and transcript data in CITE-seq experiments.
We will demonstrate using a PBMC dataset from 10X Genomics that contains quantified abundances for a number of interesting surface proteins.

```{r}
# Caching it locally with BiocFileCache to avoid repeating it.
library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)
stuff <- bfcrpath(bfc, file.path("http://cf.10xgenomics.com",
    "samples/cell-exp/3.0.0/pbmc_10k_protein_v3",
    "pbmc_10k_protein_v3_filtered_feature_bc_matrix.tar.gz"))
untar(stuff, exdir=tempdir())

# Loading it in as a SingleCellExperiment object.
library(DropletUtils)
sce <- read10xCounts(file.path(tempdir(), "filtered_feature_bc_matrix"))
sce
```

## Preprocessing 

### Setting up the data

The `SingleCellExperiment` class has an "alternative Experiment" concept to store data for different sets of features but the same cells.
This is achieved by storing another `SummarizedExperiment` (or an instance of its subclass) _inside_ our `SingleCellExperiment` where the rows can differ but the columns are the same.
In previous chapters, we were using the alternative Experiments to store spike-in data, but here we will use the concept to split off the antibody tag data.
This isolates the two sets of features to ensure that analyses on one set do not inadvertently use data from the other set, and vice versa.

```{r}
sce <- splitAltExps(sce, rowData(sce)$Type)
altExpNames(sce)
altExp(sce) # Can be used like any other SingleCellExperiment. 
```

At this point, it is also helpful to coerce the sparse matrix for antibody tags into a dense matrix.
The tag counts are usually not sparse so storage as a sparse matrix provides no advantage; in fact, it actually increases memory usage and computational time as the indices of non-zero entries must be unnecessarily stored and processed.

```{r}
counts(altExp(sce)) <- as.matrix(counts(altExp(sce)))
counts(altExp(sce))[,1:10] # sneak peek
```

### Quality control

For most part, we can depend on QC on the endogenous transcripts to remove empty droplets and low-quality cells.
Any failure in cDNA capture for a particular cell will affect the coverage of both transcripts and tags, and changes in mitochondrial RNA content are only reflected in the transcriptomic data; as a result, the tags provide little additional information for QC.
For this dataset, the count matrix has already been filtered in _CellRanger_ to remove empty droplets so we only filter on the mitochondrial proportions to remove putative low-quality cells.

```{r}
library(scater)
mito <- grep("^MT-", rowData(sce)$Symbol)
df <- perCellQCMetrics(sce, subsets=list(Mito=mito))
mito.discard <- isOutlier(df$subsets_Mito_percent, type="higher")
summary(mito.discard)
```

If we want to combine transcript and tag data, it is prudent to ensure that we only retain cells that actually have tag counts.
Recall that droplet-based libraries will contain contamination from ambient solution (Section \@ref(qc-droplets)), in this case containing containing conjugated antibodies that are either free in solution or bound to cell fragments.
As the tags are (relatively) deeply sequenced, we can expect non-zero counts for most tags in each cell (Figure \@ref(fig:detected-ab-hist); if this is not the case, we may suspect some failure of antibody tag processing for that cell.
We thus remove cells that have unusually low numbers of detected tags, defined here as half of the median across all cells.
(We could also use `isOutlier()`, but the MAD is zero in this case and filtering would discard useful cells with almost all tags detected.)

```{r detected-ab-hist, fig.cap="Distribution of the number of detected tags across all cells in the PBMC dataset."}
ab.detected <- df$`altexps_Antibody Capture_detected`
med.detected <- median(ab.detected)
threshold <- med.detected/2
hist(ab.detected, col='grey', xlab="Number of detected tags")
abline(v=threshold, col="red", lty=2)
ab.discard <- ab.detected < threshold
summary(ab.discard)
```

By comparison, the total tag count is less useful as a QC metric because it is heavily correlated with the biological state of the cell.
The presence of a targeted protein can lead to a several-fold increase in the total tag count given the binary nature of most surface protein markers.
Removing cells with low total tags would strongly penalize cell types that do not exhibit many (or any) of the selected protein targets, and would represent an extreme case of the issues discussed in Section \@ref(outlier-assumptions).

Finally, to remove the low-quality cells, we subset the `SingleCellExperiment` as previously described.
This automatically applies the filtering to both the transcript and tag data; such coordination is one of the advantages of storing both datasets in a single object.

```{r, eval=FALSE}
discard <- ab.discard | mito.discard
sce <- sce[,!discard]
```

## Normalization

## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
``` 
