---
output:
  html_document
bibliography: ../ref.bib
---

# Trajectory Analysis

```{r setup, echo=FALSE, results="asis"}
library(OSCAUtils)
chapterPreamble()
```

## Overview 

In contrast to describing population heterogenity using discrete labels (see Section \@ref(clustering-graph) for an example), some biological processes, such as cell differentiation, are better described and represented as a continuum of dynamic changes in cell states or cell types. 
This idea is often referred to as _trajectory analysis_ (or _pseudotime inference_ or lineage reconstruction) where a _trajectory_ is a potential path that a cell follows along this continuum and _pseudotime_ is the ordering induced by the this path. 
Using scRNA-seq data, the most common approaches for trajectory analysis [@saelens2019trajectory] are based on trees, probabilistic models or graphical models (see \@ref(trajectory-ideas) Section for more details), however trajectory analysis is not necessarily limited to scRNA-seq data, for example it could theoretically could be applied to epigenomic or proteomic data. 

In this section, we will walk through a minimal analysis to perform trajectory analysis and psuedotime inference for single-cells along a continuum. 
We will demonstrate this with the [`r Biocpkg("slingshot")`](https://bioconductor.org/packages/slingshot) Bioconductor package. 

We will use the 10X peripheral blood mononuclear cell (PBMC) data in the Bioconductor data package `r Biocpkg("TENxPBMCData")` that we have been using in previous chapters. 

```{r, results='asis', echo=FALSE}
extractCached("workflows/tenx-filtered-pbmc4k", "clustering", "pbmc4k")
```

```{r}
pbmc4k
```

If you recall, we created our cluster labels and have three reduced dimensions in our `r Biocpkg("SingleCellExperiment")` object. 

```{r}
head(pbmc4k$cluster)
```

```{r}
reducedDimNames(pbmc4k)
```

We will use this data in the example below, but first we will give a quick overview of the field, but refer the reader to outside references for more details [@saelens2019trajectory]. 
## Estimating pseudotime {#trajectory-ideas} 

Most trajectory methods start from a count matrix with genes along the rows and cells along the columns and then apply feature selection or a dimensionality reduction method. 
Next, these methods commonly infer pseudotime and/or branching trajectories using unsupervised clustering with a combination of minimum spanning trees (MST), principal curves (or graph fitting), or random walks and diffusion operations on graphs [@Laehnemann2019-twelvechallenges]. 

In this chapter, we will demonstrate trajectory inference with [`r Biocpkg("slingshot")`](https://bioconductor.org/packages/slingshot). 
The goal of slingshot is to use clusters of cells to uncover global structure and convert this structure into smooth lineages represented by one-dimensional variables, called "pseudotime".
The package offers tools to learn the cluster relationships in an unsupervised or semi-supervised manner and constructing smooth curves representing each lineage, along with visualization methods for each step.
Slingshot was designed to model developmental trajectories in scRNA-seq data and serve as a component in an analysis pipeline after dimensionality reduction and clustering. 

The package implements a two-step process composed of identifying the global lineage structure with a cluster-based minimum spanning tree and fitting simultaneous principal curves to describe each lineage.

These two steps can be run separately with the `getLineages()` and `getCurves()` functions, or together with the wrapper function, `slingshot()` (recommended). 
Here, we use the wrapper function for the analysis of the single-trajectory dataset. 

The `slingshot()` wrapper function performs both steps of lineage inference in a single call. 
The necessary inputs are a reduced dimensional matrix of coordinates and a set of cluster labels. 
These can be separate objects or elements contained in a `r Biocpkg("SingleCellExperiment")` object.

To run `slingshot()` with the dimensionality reduction produced by PCA and our cluster labels, we do the following:

```{r}
pbmc4k <- slingshot(pbmc4k, clusterLabels = 'cluster', reducedDim = 'PCA')
```

If no clustering results are provided, it is assumed that all cells are part of the same cluster and a single curve will be constructed. 
If no dimensionality reduction is provided, slingshot will use the first element of the list returned by `reducedDims()`.

The output is a `r Biocpkg("SingleCellExperiment")` object with `r Biocpkg("slingshot")` results incorporated. 
Most of the output is added to the metadata in the form of a list and is accessible via `metadata(pbmc4k)$slingshot`.
Additionally, all inferred pseudotime variables (one per lineage) are added to the `colData`. 
To extract all `slingshot` results in a single object, we can use the `SlingshotDataSet()` function. 
This can be useful for visualization, as several plotting methods for `r Biocpkg("SlingshotDataSet")` objects are included with the package. Below, we visuzalize the inferred lineage for the single-trajectory data with points colored by pseudotime.

```{r}
summary(pbmc4k$slingPseudotime_1)
```

We can also plot the results: 

```{r}
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(pbmc4k$slingPseudotime_1, breaks=100)]

plot(reducedDims(pbmc4k)$PCA, col = plotcol, pch=16, asp = 1)
lines(SlingshotDataSet(pbmc4k), lwd=2, col='black')
```

We can also see how the lineage structure was intially estimated by the cluster-based minimum spanning tree by using the type argument.

```{r}
plot(reducedDim(pbmc4k, "PCA"), col = brewer.pal(9,'Set1')[pbmc4k$cluster], pch=16, asp = 1)
lines(SlingshotDataSet(pbmc4k), lwd=2, type = 'lineages', col = 'black')
```

## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
```
