---
output:
  html_document
bibliography: ../ref.bib
---

# Trajectory Analysis

```{r setup, echo=FALSE, results="asis"}
library(OSCAUtils)
chapterPreamble()
```

## Overview 

In contrast to describing population heterogenity using discrete labels (see Section \@ref(clustering-graph) for an example), some biological processes, such as cell differentiation, are better described and represented as a continuum of dynamic changes in cell states or cell types. 
This idea is often referred to as _trajectory analysis_ (or _pseudotime inference_ or lineage reconstruction) where a _trajectory_ is a potential path that a cell follows along this continuum and _pseudotime_ is the ordering induced by the this path. 
Using scRNA-seq data, the most common approaches for trajectory analysis [@saelens2019trajectory] are based on trees, probabilistic models or graphical models (see \@ref(trajectory-ideas) Section for more details), however trajectory analysis is not necessarily limited to scRNA-seq data, for example it could theoretically could be applied to epigenomic or proteomic data. 

In this section, we will walk through a minimal analysis to perform trajectory analysis and psuedotime inference for single-cells along a continuum. 
We will demonstrate this with the [`r Biocpkg("slingshot")`](https://bioconductor.org/packages/slingshot) Bioconductor package. 

We will use the 10X peripheral blood mononuclear cell (PBMC) data in the Bioconductor data package `r Biocpkg("TENxPBMCData")` that we have been using in previous chapters. 

```{r, results='asis', echo=FALSE}
extractCached("workflows/tenx-filtered-pbmc4k", "clustering", "pbmc4k")
```

```{r}
pbmc4k
```

If you recall, we created our cluster labels and have three reduced dimensions in our `r Biocpkg("SingleCellExperiment")` object. 

```{r}
head(pbmc4k$cluster)
```

```{r}
reducedDimNames(pbmc4k)
```

We will use this data in the example below, but first we will give a quick overview of the field, but refer the reader to outside references for more details [@saelens2019trajectory]. 

## Estimating pseudotime {#trajectory-ideas} 

Most trajectory methods start from a count matrix with genes along the rows and cells along the columns and then apply feature selection or a dimensionality reduction method. 
Next, these methods commonly infer pseudotime and/or branching trajectories using unsupervised clustering with a combination of minimum spanning trees (MST), principal curves (or graph fitting), or random walks and diffusion operations on graphs [@Laehnemann2019-twelvechallenges]. 
In the next few sections, we will discuss these ideas briefly, but much of this discussion comes from Street et al. [@Street2018-slingshot]. 

### Minimum spanning trees 

One of the most well-known approaches is to construct a minimum spanning tree (MST) on cells in a reduced-dimensionality space. 
For example, the Monocle [@Trapnell2014] package, constructs a MST on cells after applying independent component analysis (ICA) and orders cells along the longest path through this tree.
The number of branching events and the direction of the path is left to the user, but this can be explored using a known set of marker genes or use time of sample collection as indications of initial and terminal cell states. 
The more recent Monocle 2 [@qiu2017census] package uses a different approach, namely it applies dimensionality reduction and ordering performed by reverse graph embedding (RGE), allowing it to detect branching events in an unsupervised manner. 
Alternatively, the methods Waterfall [@Shin2015-waterfall] and TSCAN [@Ji2016-tscan] instead determine the lineage structure by clustering cells in a low-dimensional space and drawing an MST on the cluster centers. 
Lineages are represented by piecewise linear paths through the tree, providing an intuitive, unsupervised method for identifying branching events. 
Pseudotimes are calculated by orthogonal projection onto these paths, with the identification of the direction and of the cluster of origin again left to the user. 

### Cell-to-cell distances 

Other approaches use robust cell-to-cell distances and a pre-specified starting cell to determine pseudotime. 
For instance, diffusion pseudotime (DPT) [@Haghverdi2016-dpt] uses a weighted _k_ nearest neighbors (kNN) graph on cells and calculates distances using transition probabilities over random walks of arbitrary length. 
Similarly, Wishbone [@Setty2016-wishbone] uses an ensemble of kNN graphs on cells along with a randomly selected group of waypoints to iteratively refine stable distance estimates. 

### Model-based approaches

Model-based approaches can be used to detect branching events. 
GPfates [@Lonnberg-gpfates] uses a Gaussian process latent variable model (GPLVM) and overlapping mixtures of Gaussian processes (OMGP) to infer trajectories and pseudotimes. 
A similar method, DeLorean [@Reid2016-delorean], uses a single GPLVM to infer pseudotimes along a single trajectory. 
And the mixtures of factor analysers (MFA) method [@Campbell2017-MFA] takes a hierarchical Bayesian approach, using Markov chain Monte-Carlo (MCMC) to sample from the posterior of a fully generative model that includes branch identities. 

### Principal curves

Other approaches use smooth curves to represent development, but are naturally limited to non- branching lineages. 
For example, Embeddr uses the principal curves to infer lineages in a low-dimensional space obtained by a Laplacian eigenmap. 
However, slingshot [@Street2018-slingshot] leverages _simultaneous principal curves_ to allow for multiple branching lineages.

In the next section, we will demonstrate trajectory inference with [`r Biocpkg("slingshot")`](https://bioconductor.org/packages/slingshot). 
The goal of slingshot is to use clusters of cells to uncover global structure and convert this structure into smooth lineages represented by one-dimensional variables, called "pseudotime".
The package offers tools to learn the cluster relationships in an unsupervised or semi-supervised manner and constructing smooth curves representing each lineage, along with visualization methods for each step.
Slingshot was designed to model developmental trajectories in scRNA-seq data and serve as a component in an analysis pipeline after dimensionality reduction and clustering. 

#### Implementation 

The package implements a two-step process composed of identifying the global lineage structure with a cluster-based MST and fitting simultaneous principal curves to describe each lineage.

In the first step, the advantage of using a cluster-based MST is to stably identify the key elements of the global lineage structure, i.e., the number of lineages and where they branch.
This allows slingshot to identify novel lineages while also accommodating the use of domain-specific knowledge to supervise parts of the tree (e.g., terminal cellular states). 
In the second step, the simultaneous principal curves allow for the translation the knowledge of global lineage structure into stable estimates of the underlying cell-level pseudotime variable for each lineage. 

These two steps can be run separately with the `getLineages()` and `getCurves()` functions, or together with the wrapper function, `slingshot()` (recommended). 
Here, we use the wrapper function for the analysis of the single-trajectory dataset. 

The `slingshot()` wrapper function performs both steps of lineage inference in a single call. 
The necessary inputs are a reduced dimensional matrix of coordinates and a set of cluster labels. 
These can be separate objects or elements contained in a `r Biocpkg("SingleCellExperiment")` object.

To run `slingshot()` with the dimensionality reduction produced by PCA and our cluster labels, we do the following:

```{r}
pbmc4k <- slingshot(pbmc4k, clusterLabels = 'cluster', reducedDim = 'PCA')
```

If no clustering results are provided, it is assumed that all cells are part of the same cluster and a single curve will be constructed. 
If no dimensionality reduction is provided, slingshot will use the first element of the list returned by `reducedDims()`.

The output is a `r Biocpkg("SingleCellExperiment")` object with `r Biocpkg("slingshot")` results incorporated. 
Most of the output is added to the metadata in the form of a list and is accessible via `metadata(pbmc4k)$slingshot`.
Additionally, all inferred pseudotime variables (one per lineage) are added to the `colData`. 
To extract all `slingshot` results in a single object, we can use the `SlingshotDataSet()` function. 
This can be useful for visualization, as several plotting methods for `r Biocpkg("SlingshotDataSet")` objects are included with the package. Below, we visuzalize the inferred lineage for the single-trajectory data with points colored by pseudotime.

```{r}
summary(pbmc4k$slingPseudotime_1)
```

We can also plot the results: 

```{r}
colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(pbmc4k$slingPseudotime_1, breaks=100)]

plot(reducedDims(pbmc4k)$PCA, col = plotcol, pch=16, asp = 1)
lines(SlingshotDataSet(pbmc4k), lwd=2, col='black')
```

We can also see how the lineage structure was intially estimated by the cluster-based minimum spanning tree by using the type argument.

```{r}
plot(reducedDim(pbmc4k, "PCA"), col = brewer.pal(9,'Set1')[pbmc4k$cluster], pch=16, asp = 1)
lines(SlingshotDataSet(pbmc4k), lwd=2, type = 'lineages', col = 'black')
```

## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
prettySessionInfo()
```
